flink-table-store-core/src/main/java/org/apache/flink/table/store/file/mergetree/compact/FullChangelogMergeFunctionWrapper.java

```diff
@@ -0,0 +1,183 @@
+        initialKv = null;
+        isInitialized = false;
+    }
+
+    @Override
+    public void add(KeyValue kv) {
+        if (maxLevel == kv.level()) {
+            Preconditions.checkState(
+                    topLevelKv == null, "Top level key-value already exists! This is unexpected.");
+            topLevelKv = kv;
+        }
+
+        if (initialKv == null) {
+            initialKv = kv;
+        } else {
+            if (!isInitialized) {
+                merge(initialKv);
+                isInitialized = true;
+            }
+            merge(kv);
+        }
+    }
+
+    private void merge(KeyValue kv) {
+        mergeFunction.add(kv);
+    }
+
+    @Override
+    public Result getResult() {
+        if (reusedResult == null) {
+            reusedResult = new Result();
+            reusedBefore = new KeyValue();
+            reusedAfter = new KeyValue();
+        }
+
+        if (isInitialized) {
+            KeyValue merged = mergeFunction.getResult();
+            if (topLevelKv == null) {
+                if (merged != null && isAdd(merged)) {
+                    reusedResult.setChangelog(null, replace(reusedAfter, RowKind.INSERT, merged));
+                } else {
+                    reusedResult.setChangelog(null, null);
+                }
+            } else {
+                if (merged != null && isAdd(merged)) {
+                    reusedResult.setChangelog(
+                            replace(reusedBefore, RowKind.UPDATE_BEFORE, topLevelKv),
+                            replace(reusedAfter, RowKind.UPDATE_AFTER, merged));
+                } else {
+                    reusedResult.setChangelog(
+                            replace(reusedBefore, RowKind.DELETE, topLevelKv), null);
+                }
+            }
+            return reusedResult.setResult(merged);
+        } else {
+            if (topLevelKv == null && isAdd(initialKv)) {
+                reusedResult.setChangelog(null, replace(reusedAfter, RowKind.INSERT, initialKv));
+            } else {
+                // either topLevelKv is not null, but there is only one kv,
+                // so topLevelKv must be the only kv, which means there is no change
+                //
+                // or initialKv is not an ADD kv, so no new key is added
+                reusedResult.setChangelog(null, null);
+            }
+            return reusedResult.setResult(initialKv);
+        }
+    }
+
+    private KeyValue replace(KeyValue reused, RowKind valueKind, KeyValue from) {
+        return reused.replace(from.key(), from.sequenceNumber(), valueKind, from.value());
+    }
+
+    private boolean isAdd(KeyValue kv) {
+        return kv.valueKind() == RowKind.INSERT || kv.valueKind() == RowKind.UPDATE_AFTER;
+    }
+
+    @Override
+    public MergeFunction<Result> copy() {
+        return new FullChangelogMergeFunctionWrapper(mergeFunction, maxLevel);
+    }
+
+    /** Changelog and final result for the same key. */
+    public static class Result {
```

Comment:
Should we extract a separate class for this? (Not a inner class)

Reply:
No need, at least currently. This result is only used for producing changelogs during full compaction. If it will be widely used afterwards then we can consider making it a separate class.